### ISSUE TITLE
Multiple `setup_rnn` calls

### ISSUE THREAD

**Issue by kradonneoh** [HazyResearch/state-spaces#70]
Hey!

I'm attempting to integrate the Sashimi Backbone into some audio models -- I'd like to train in convolutional mode and run _validation_ inference in RNN mode, but my reading of the code seems to imply that the `setup_step` call isn't repeatable or reversible (#67 seems to imply this as well).

In the case that I temporarily want to infer in RNN mode, but then switch back to the convolutional training mode, what's my best option?

**Comment by albertfgu** [comment]
I think it should actually work fine calling it multiple times. You will just need to call it before every time you want to use the RNN mode.

**Comment by kradonneoh** [comment]
Ah, I must've misread the code / the paper -- when we call `setup_rnn`, the kernel state matrices are initialized for step calls only, but not for regular forward calls? In that, there's no leakage into the convolutional mode?

**Comment by kradonneoh** [comment]
While I have you, I'm attempting to understand the output of the Sashimi forward pass and how I'd do a simple cross entropy loss on the predictions. If I'm reading the Sashimi backbone right, in the UpPool blocks, we shift the inputs one over to the right (by padding left and removing the last element). Does that imply, if `y = self.forward(x)`, the `i`th element of `y` is the prediction based on the `< i` elements of `x`?

**Comment by kradonneoh** [comment]
Sweet, that lines up with my intuition. One last question for you: when trying to generate _conditionally_, my thought was to convert to RNN, initialize default state, then feed the conditioning sequence through the model to build up state non-auto-regressively (e.g. predict `x_i` given `x_i-1`, discarding output `y_i`s), then auto-regressively generate the output sequence (e.g. predict `y_i` given `y_i-1`)

Does this line up with how one conditions S4 based models?

**Comment by kradonneoh** [comment]
Thank you, appreciate the quick response times.
